harness/determined/exec/harness.py

```diff
@@ -129,21 +130,22 @@ def build_and_run_training_pipeline(env: det.EnvContext) -> None:
             # Horovod distributed training is done inside subprocesses.
             if hvd_config.use:
                 subproc = layers.SubprocessLauncher(
-                    env, iter(workload_mgr), load_path, socket_mgr.get_rendezvous_info(), hvd_config
+                    env, workloads, load_path, socket_mgr.get_rendezvous_info(), hvd_config
                 )
                 subproc.run()
             else:
                 if env.experiment_config.debug_enabled():
                     faulthandler.dump_traceback_later(30, repeat=True)
 
                 with det._catch_sys_exit():
-                    controller = load.prepare_controller(
-                        env,
-                        iter(workload_mgr),
-                        load_path,
-                        socket_mgr.get_rendezvous_info(),
-                        hvd_config,
-                    )
+                    with det._catch_init_invalid_hp(workloads):
```

Comment:
If you catch the exception here, you will only be able to catch when executing users' trial init function.

Reply:
This is the intended functionality for this context manager. We will handle invalid hp exceptions raised in training and validations workloads in the individual trial controllers.

Reply:
It might be worth mentioning why it is designed this way (I helped with this design haha).  This is ultimately a way for us to hack around the layers abstraction of the harness, which in some cases needs to say alive long enough for the master to ask for a checkpoint (right?).

But if there's a failure at init time, we don't need to do that, and we can just exit out of the harness.

(if we could push checkpoints, then we could always decide if we want a checkpoint at the moment we catch the error, then push a checkpoint, then exit, and things would be a bit simpler)

Reply:
I have added a checkpoint API in [a recent PR](https://github.com/determined-ai/determined/pull/1968). Do you think it would help this PR to get less complicated? @rb-determined-ai

Reply:
absolutely.